File: android/app/src/main/kotlin/org/gaindrive/android/data/model/Library.kt
package org.gaindrive.android.data.model

/**
 * Domain models. Every one carries an [ItemRef] rather than a bare id, so the
 * server an item came from travels with it — which is what makes a merged
 * library, a mixed-server queue and per-item actions possible at all.
 *
 * Separate from the DTOs on purpose: Phase 6's cache will produce these from
 * Room as well as from the network, and retrofitting that split later would
 * touch every screen.
 */

data class Artist(
	val ref: ItemRef,
	val name: String,
	val albumCount: Int,
	val coverArt: ItemRef?,
	val starredAt: String?,
	/**
	 * Every (server, id) pair this row stands for. One entry normally; several
	 * when artists of the same name have been merged across servers, in which
	 * case [ref] is the first contributor in registry order.
	 *
	 * Ids, not just server ids: opening a merged artist has to ask each server
	 * for *its* artist, and the ids differ.
	 */
	val refs: List<ItemRef> = listOf(ref),
) {
	val isStarred: Boolean get() = starredAt != null

	/** The servers contributing to this row, for the badges. */
	val sources: List<ServerId> get() = refs.map { it.server }
}

/**
 * One configured library root, from `getMusicFolders`.
 *
 * [contentType] is a gaindrive extension and is null everywhere else — which is
 * the whole of what tells a server that knows about kinds of root from one that
 * only has folders. See `data/browse/LibraryRoots.kt`.
 */
data class MusicRoot(
	val id: String,
	val name: String,
	val contentType: String?,
)

/** One index bucket from `getArtists`, e.g. "S" or "#". */
data class ArtistIndex(
	val label: String,
	val artists: List<Artist>,
)

data class Album(
	val ref: ItemRef,
	val title: String,
	val artistName: String,
	val artistRef: ItemRef?,
	val songCount: Int,
	val duration: Int,
	val year: Int?,
	val genre: String?,
	val coverArt: ItemRef?,
	val starredAt: String?,
	/**
	 * Every server holding this album. One entry unless duplicates have been
	 * collapsed, in which case [ref] is the copy that won — the one highest in
	 * registry order — and the rest are kept so the row can still say who else
	 * has it.
	 */
	val refs: List<ItemRef> = listOf(ref),
) {
	val isStarred: Boolean get() = starredAt != null

	/** The servers holding this album, for the badges. */
	val sources: List<ServerId> get() = refs.map { it.server }
}

data class Song(
	val ref: ItemRef,
	val title: String,
	/** Who made this track, which on a compilation is not [albumArtistName]. */
	val artistName: String,
	/**
	 * Who the album is by. Blank when the server did not say — an older one, or
	 * a listing built offline before this was mirrored — which is why
	 * [differingArtist] insists on having it before drawing anything.
	 */
	val albumArtistName: String = "",
	val albumTitle: String,
	val albumRef: ItemRef?,
	val track: Int?,
	val discNumber: Int?,
	val year: Int?,
	val duration: Int,
	val bitRate: Int,
	val suffix: String?,
	val contentType: String?,
	val sizeBytes: Long,
	val coverArt: ItemRef?,
	val starredAt: String?,
	/** Only populated by the recents query. */
	val lastPlayedAt: String? = null,
	val isVideo: Boolean = false,
	/**
	 * Whether the stream this song would produce can be seeked by byte range.
	 * False for a video the server can only re-encode on the fly, which is what
	 * sends playback to `hls.m3u8` instead. Meaningless for audio.
	 */
	val nativeSeek: Boolean = false,
	/**
	 * The season an episode belongs to, or null. [discNumber] holds the same
	 * number and is what orders and groups tracks; this only decides whether
	 * the group is headed "Series 2" or "Disc 2".
	 */
	val season: Int? = null,
	/**
	 * What the server will actually send if it has to convert this, or null when
	 * it will send the file as it stands.
	 *
	 * Only casting reads it, and it is the reason casting does not have to work
	 * out the tier for itself: the receiver picks its decode pipeline from the
	 * declared type, and for a video the *source* `contentType` is wrong in
	 * exactly the common case — an H.264/AAC `.mkv` is remuxed and arrives as
	 * `video/mp4`, not as `video/x-matroska`.
	 */
	val transcodedContentType: String? = null,
	/** Video frame size, when the server could probe it. */
	val width: Int? = null,
	val height: Int? = null,
) {
	val isStarred: Boolean get() = starredAt != null

	/**
	 * The track's own artist when the album is not by them, and null otherwise
	 * — so a row can draw it without deciding anything.
	 *
	 * An exact comparison on purpose. The server already answered the hard
	 * half: it sends the folder's spelling in [artistName] whenever the file's
	 * tag is merely a different way of writing the same name, so "AC/DC"
	 * against a folder called "AC-DC" never reaches here as a difference.
	 */
	val differingArtist: String?
		get() = artistName.takeIf {
			it.isNotBlank() && albumArtistName.isNotBlank() && it != albumArtistName
		}

	/** The frame's aspect, or null when the dimensions are unknown. */
	val aspectRatio: Float?
		get() = if (width != null && height != null && width > 0 && height > 0) {
			width.toFloat() / height.toFloat()
		} else {
			null
		}
}

data class Playlist(
	val ref: ItemRef,
	val name: String,
	val comment: String?,
	val owner: String?,
	val isPublic: Boolean,
	val songCount: Int,
	val duration: Int,
	val songs: List<Song> = emptyList(),
)

/** Biography and outbound links for an artist, from `getArtistInfo2`. */
data class ArtistInfo(
	val biography: String?,
	val wikiUrl: String?,
	val allMusicUrl: String?,
	val lastFmUrl: String?,
	val discogsUrl: String?,
) {
	val isEmpty: Boolean
		get() = biography.isNullOrBlank() && wikiUrl.isNullOrBlank() &&
			allMusicUrl.isNullOrBlank() && lastFmUrl.isNullOrBlank() &&
			discogsUrl.isNullOrBlank()
}

/** Notes and outbound links for an album, from `getAlbumInfo2`. */
data class AlbumNotes(
	val notes: String?,
	val wikiUrl: String?,
	val allMusicUrl: String?,
) {
	val isEmpty: Boolean
		get() = notes.isNullOrBlank() && wikiUrl.isNullOrBlank() && allMusicUrl.isNullOrBlank()
}

/**
 * An album and its tracks — the part of the detail screen that is worth
 * blocking on. The notes are fetched separately (see [AlbumNotes]) because
 * they can be slow and are never essential.
 */
data class AlbumDetail(
	val album: Album,
	val songs: List<Song>,
)

/** Grouped results from a search or from the starred list. */
data class LibrarySelection(
	val artists: List<Artist> = emptyList(),
	val albums: List<Album> = emptyList(),
	val songs: List<Song> = emptyList(),
	/**
	 * Chapter markers whose titles matched, which only a search fills in.
	 *
	 * Defaulted because the starred list shares this type and never has any,
	 * and because nothing mirrors them: a marker has no id Room could key on,
	 * so the offline branch and the stored fallback both leave this empty and
	 * the Chapters section simply does not appear.
	 */
	val chapters: List<ChapterHit> = emptyList(),
) {
	val isEmpty: Boolean
		get() = artists.isEmpty() && albums.isEmpty() && songs.isEmpty() && chapters.isEmpty()
}

/** What `star`/`unstar` is being applied to; each takes a different parameter. */
enum class StarKind { SONG, ALBUM, ARTIST }

File: ios/GainDrive/Data/Model/Library.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

//	The domain models, mirroring `data/model/Library.kt`.
//
//	Separate from the DTOs on purpose. It is tempting to skip the mapping layer
//	at this size, but phase 5's offline mirror needs models that can come from
//	either the network or the local database, and retrofitting that split later
//	would touch every screen.
//
//	Every identifier here is an `ItemRef`. Nothing in this file holds a bare
//	`String` id, and nothing above it should either.
//
//	They are `Codable` for the mirror, which stores mapped domain values rather
//	than raw responses — see `LibraryMirror`. That is also the whole of the
//	mirror's migration story: a value written by a build with a different shape
//	fails to decode, which reads as a miss, which is a re-fetch. A cache is
//	allowed to be thrown away.

/// One of a server's configured library roots.
///
/// **`id` is a bare `String`, and that is not a lapse in the composite-id
/// rule.** A root id is only meaningful to the server that issued it, and this
/// one never travels: it is read from that server's `getMusicFolders` and put
/// straight back into a request to the same server. Nothing above `Data` ever
/// sees it — the merged listing is asked for by root *kind*
/// (`LibraryRoots.listingRequests`), never by a root's own id.
struct MusicRoot: Hashable, Sendable, Codable {
	let id: String
	let name: String
	/// A gaindrive extension: `artists` or `categories`. Absent on a server
	/// with no concept of root kinds — which is **not** the same as having no
	/// roots of that kind.
	let contentType: String?
}

/// An artist, possibly standing for the same artist on several servers.
struct Artist: Identifiable, Hashable, Sendable, Codable {
	let ref: ItemRef
	let name: String
	let albumCount: Int
	let coverArt: ItemRef?
	let starredAt: String?
	/// **Ids, not just server ids.** Opening a merged artist has to ask each
	/// contributing server for *its* artist, and those ids differ. A row that
	/// remembered only which servers it came from could not do that.
	let refs: [ItemRef]

	init(
		ref: ItemRef, name: String, albumCount: Int, coverArt: ItemRef? = nil,
		starredAt: String? = nil, refs: [ItemRef]? = nil
	) {
		self.ref = ref
		self.name = name
		self.albumCount = albumCount
		self.coverArt = coverArt
		self.starredAt = starredAt
		self.refs = refs ?? [ref]
	}

	var id: ItemRef { ref }
	var isStarred: Bool { starredAt != nil }
	var sources: [ServerId] { refs.map(\.server) }
}

/// One bucket of the index rail: a letter and the artists filed under it.
struct ArtistIndex: Identifiable, Hashable, Sendable, Codable {
	let label: String
	let artists: [Artist]

	var id: String { label }
}

struct Album: Identifiable, Hashable, Sendable, Codable {
	let ref: ItemRef
	let title: String
	let artistName: String
	let artistRef: ItemRef?
	let songCount: Int
	/// How many of this album's tracks are video, which is the only thing that
	/// tells a season or a film from a record before its tracks are fetched —
	/// `isVideo` is a per-song field.
	///
	/// **Zero does not mean "no video".** The server does not carry it on the
	/// directory-shaped listings, so an album reached through search or starred
	/// reports zero whatever it holds. That is why the mark is only ever added
	/// and there is no "audio" counterpart: a wrong positive would be a lie,
	/// and a missing one is silence.
	let videoCount: Int
	let duration: Int
	let year: Int?
	let genre: String?
	let coverArt: ItemRef?
	let starredAt: String?
	/// As `Artist.refs` — a merged album row stands for the same record on
	/// several servers.
	let refs: [ItemRef]

	init(
		ref: ItemRef, title: String, artistName: String, artistRef: ItemRef? = nil,
		songCount: Int = 0, videoCount: Int = 0, duration: Int = 0, year: Int? = nil,
		genre: String? = nil, coverArt: ItemRef? = nil, starredAt: String? = nil,
		refs: [ItemRef]? = nil
	) {
		self.ref = ref
		self.title = title
		self.artistName = artistName
		self.artistRef = artistRef
		self.songCount = songCount
		self.videoCount = videoCount
		self.duration = duration
		self.year = year
		self.genre = genre
		self.coverArt = coverArt
		self.starredAt = starredAt
		self.refs = refs ?? [ref]
	}

	var id: ItemRef { ref }
	var isStarred: Bool { starredAt != nil }
	var sources: [ServerId] { refs.map(\.server) }
}

/// A track. **Songs never merge** — a track list always comes from one album on
/// one server — so there is no `refs` here, and its absence is the rule.
struct Song: Identifiable, Hashable, Sendable, Codable {
	let ref: ItemRef
	let title: String
	let artistName: String
	let albumTitle: String
	let albumRef: ItemRef?
	let track: Int?
	let discNumber: Int?
	let year: Int?
	let duration: Int
	let bitRate: Int?
	let suffix: String?
	let contentType: String?
	let sizeBytes: Int
	let coverArt: ItemRef?
	let starredAt: String?
	let lastPlayedAt: String?
	let isVideo: Bool
	/// False whenever the endpoint did not select the codec columns, which is
	/// the safe direction — the video still plays, it just seeks by
	/// re-request. Trusted as given rather than second-guessed.
	let nativeSeek: Bool
	let width: Int?
	let height: Int?

	var id: ItemRef { ref }
	var isStarred: Bool { starredAt != nil }
}

struct Playlist: Identifiable, Hashable, Sendable, Codable {
	let ref: ItemRef
	let name: String
	let comment: String?
	let owner: String?
	let isPublic: Bool
	let songCount: Int
	let duration: Int
	let songs: [Song]

	var id: ItemRef { ref }
}

struct ArtistInfo: Hashable, Sendable {
	let biography: String?
	let wikiUrl: String?
	let allMusicUrl: String?
	let lastFmUrl: String?
	let discogsUrl: String?
	let imageUrl: String?

	/// Worth asking before rendering a header that would otherwise be an empty
	/// box: the server answers `ok` with nothing in it for an artist
	/// MusicBrainz has never heard of.
	var isEmpty: Bool {
		(biography?.isEmpty ?? true) && wikiUrl == nil && allMusicUrl == nil
			&& lastFmUrl == nil && discogsUrl == nil
	}
}

struct AlbumNotes: Hashable, Sendable {
	let notes: String?
	let wikiUrl: String?
	let allMusicUrl: String?

	var isEmpty: Bool {
		(notes?.isEmpty ?? true) && wikiUrl == nil && allMusicUrl == nil
	}
}

struct AlbumDetail: Hashable, Sendable, Codable {
	let album: Album
	let songs: [Song]

	/// Disc headings appear only when there is more than one disc, so a
	/// single-disc album is not decorated with a heading that says nothing.
	var isMultiDisc: Bool {
		Set(songs.compactMap(\.discNumber)).count > 1
	}
}

/// The shape both `search3` and `getStarred2` answer with.
struct LibrarySelection: Hashable, Sendable {
	var artists: [Artist] = []
	var albums: [Album] = []
	var songs: [Song] = []
	/// **A list of its own, never entries among the songs**, which is how the
	/// server sends it and for its reason: a chapter has no id anything can
	/// stream, star or queue, so a client told it was a song would be handed a
	/// track that does not work.
	var chapters: [ChapterHit] = []

	var isEmpty: Bool {
		artists.isEmpty && albums.isEmpty && songs.isEmpty && chapters.isEmpty
	}
}

/// Which parameter name `star`/`unstar` should carry the id under.
enum StarKind: Sendable, Hashable {
	case song, album, artist
}
